### ISSUE TITLE
How to log messages and error so that they appear in the form?

### ISSUE THREAD

**Issue by iSapozhnik** [MacPaw/FMPFeedbackForm#18]
Hey and thanks for the great form. One question though... I'm using `Logger` to log some messages and errors but this does not propagate data into the form. Am I'm doing something wrong?

Here is what I'm using:

```swift
import os

enum Log {
    private static let logger = Logger()
    
    static func notice(_ message: String) {
        logger.notice("ℹ️ \(message)")
    }
    
    static func error(_ message: String) {
        logger.error("🚨 \(message)")
    }
    
    static func warning(_ message: String) {
        logger.warning("⚠️ \(message)")
    }
}
```

But the `Console` tab is empty in the form...
![image](https://user-images.githubusercontent.com/1089384/199232940-97d88f39-316d-4b76-b997-c2c0962e4aa6.png)

**Comment by anton-barkov** [comment]
Hi @iSapozhnik, glad you like the form :)

Does this Logger write logs into a text file? If so, you must explicitly specify a path to this file to the form's controller so that it can be included in the system report. Like so: 
```swift
feedbackFormController.logURLs = [URL(fileURLWithPath: "path/to/file.log")]
```

Please see the corresponding section in the doc for more info:
https://github.com/MacPaw/FMPFeedbackForm#system-profile-report-1

**Comment by iSapozhnik** [comment]
Hey Anton, thanks for the reply. I'm not sure if it logs into a file. But at least I see my messages in the Console app. It's a native logger from Apple which is just a fancy wrapper around os.log [link](https://developer.apple.com/documentation/os/logger)

**Comment by anton-barkov** [comment]
Hm, I see, haven't used it yet.
At a glance, I don't see a way to read the messages from some kind of Logger's storage. If the messages didn't show up by themselves, I'm afraid there's no quick fix for this approach.

If you want a surefire solution though, you may try using a 3rd party logging tool that writes log messages into a file. You'll then specify this file's URL to the form.

For example, you may try [SwiftyBeaver](https://github.com/SwiftyBeaver/SwiftyBeaver) - it has a handy static interface that looks a lot like the snippet you've provided above. There shouldn't be a lot of refactoring involved if you decide to switch to this tool.

**Comment by iSapozhnik** [comment]
Yeah ok, I will investigate how to collect logs and if I find a way I will just collect them manually, save them to a file and then feed them into the form.

And the reason why my snippet looks pretty much like swiftybeaver is that I've used it before in my other project but this time I decided to keep my dependency tree cleaner. But if the previous approach failed I will switch to it again. Thanks for the support and a great form.

**Comment by iSapozhnik** [comment]
Just in case someone would need the same feature...
Here is my implementation of the idea mentioned above.

```swift
import OSLog

/// Used to log messages to the console
enum Log {
    private static let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "main")
    
    static func debug(_ message: String) {
        logger.debug("🛠 \(message)")
    }
    
    static func notice(_ message: String) {
        logger.notice("ℹ️ \(message)")
    }
    
    static func error(_ message: String) {
        logger.error("🚨 \(message)")
    }
    
    static func error(_ error: Error) {
        logger.error("🚨 \(error.localizedDescription)")
    }
    
    static func warning(_ message: String) {
        logger.warning("⚠️ \(message)")
    }
}

extension Log {
    struct LogEntry: Codable {
        let level: Int
        let subsystem, category, composedMessage: String
        let date: Date
        
        var formattedString: String {
            "[\(level)] [\(date.formatted())] [\(category)] [\(subsystem)] \(composedMessage)"
        }
    }

    struct LogData: Codable {
        let url: URL
        let logEntries: [LogEntry]
    }
    
    /// Is used to get the log data from the device
    /// - parameter completion: `(Result<LogData, Error>) -> Void` A closure that
    /// takes a `Result<LogData, Error>` and returns nothing
    static func logData(_ completion: @escaping (Result<LogData, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let logData = try prepareLogData()
                DispatchQueue.main.async {
                    completion(.success(logData))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
    
    private static func prepareLogData() throws -> LogData {
        let url = makeLogURL()
        
        let logEntries = try logEntries()
        let logEntriesString = logEntries.map(\.formattedString).joined(separator: "\n")
        try logEntriesString.write(to: url, atomically: true, encoding: .utf8)
        notice("Logs generated at: \(url.absoluteString)")
        return LogData(url: url, logEntries: logEntries)
    }
    
    private static func logEntries() throws -> [LogEntry] {
        let allEntries = try getLogEntries()
        return allEntries.map { LogEntry(level: $0.level.rawValue,
                                         subsystem: $0.subsystem,
                                         category: $0.category,
                                         composedMessage: $0.composedMessage,
                                         date: $0.date)
        }
    }
    
    private static func getLogEntries() throws -> [OSLogEntryLog] {
        let logStore = try OSLogStore(scope: .system)
        let oneHourAgo = logStore.position(date: Date.now.addingTimeInterval(-60 * 60))
        let subsystemPredicate = NSPredicate(format: "subsystem == %@ AND category == %@",
                                             argumentArray: [Bundle.main.bundleIdentifier!, "main"])
        let allEntries = try logStore.getEntries(with: [], at: oneHourAgo, matching: subsystemPredicate)
        let subsystem = Bundle.main.bundleIdentifier!
        return allEntries
            .compactMap { $0 as? OSLogEntryLog }
            .filter { $0.subsystem == subsystem } // This line might not even be needed since we are using predicates
    }
    
    private static func makeLogURL() -> URL {
        let fileName = "\(UUID().uuidString).log"
        let url = FileManager.default.temporaryDirectory
        if #available(macOS 13.0, *) {
            return url.appending(path: fileName)
        } else {
            return url.appendingPathComponent(fileName)
        }
    }
}
```

How to use it:
```swift
Log.logData { result in
    switch result {
    case let .success(logData):
        self.feedbackController?.logURLs = [logData.url]
    case let .failure(error):
        Log.error(error.localizedDescription)
    }
}
```
